### PR TITLE
feat:  add k8s subcommand

### LINKED ISSUE #2062: Add sub command for k8s
## Description
Add sub-command for k8s, and use kubernetes-trivy lib.

### PR DESCRIPTION
**Pull request by josedonizetti**
## Description

Add k8s subcommand

## Related issues
- Close #2062 

## Related PRs
- https://github.com/aquasecurity/trivy-kubernetes/pull/6

Remove this section if you don't have related PRs.

## Checklist
- [ ] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [ ] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [ ] I've added tests that prove my fix is effective or that my feature works.
- [ ] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

**Comment by knqyf263** [comment]
I'm curious why we needed another PR.
https://github.com/aquasecurity/trivy/pull/2055

**Comment by josedonizetti** [comment]
@knqyf263 Sorry, I was working on my fork, endup branching from you and creating a new branch to in case I broke something on your branch. I'll close this PR and push my changes to your branch/PR.

File: pkg/commands/artifact/k8s.go
Comment (chen-keinan) [inline pkg/commands/artifact/k8s.go:163]: Maybe the logic for persisting files on fs should be done in `trivy-kubernetes`, shouldn't trivy accept only root file folder path  and walk it?

File: pkg/commands/artifact/k8s.go
Comment (chen-keinan) [inline pkg/commands/artifact/k8s.go:166]: we are calling k8s scan for each file path, shouldn't we call it once for folder path?

**Comment by knqyf263** [comment]
>@knqyf263 Sorry, I was working on my fork, endup branching from you and creating a new branch in case I broke something on your branch and forgot about your PR. I'll close this PR and push my changes to your branch/PR.

I'll close my PR then. I just wanted to know why we needed two PRs. If it is just a mistake, it is ok👍

File: pkg/commands/option/kubernetes.go
Comment (knqyf263) [inline pkg/commands/option/kubernetes.go:23]: If there is no validation, we don't have to have `Init()`.

File: go.mod
Comment (knqyf263) [inline go.mod:221]: Could you point to the existing commit in remote so that reviewers can run this PR easily? Like `go get -u github.com/aquasecurity/trivy-kubernetes@fcfb51c0e4ebb6941377f1a4e732de54a318847f`. After merging https://github.com/aquasecurity/trivy-kubernetes/pull/6, we will update go.mod again to point to the main branch. I don't mean you should not use `replace` directive. You can use `replace` in local, but it would be helpful if you use the actual commit when pushing commits to GitHub.

File: pkg/commands/artifact/k8s.go
Comment (knqyf263) [inline pkg/commands/artifact/k8s.go:145]: ioutil is deprecated

File: go.mod
Comment (josedonizetti) [reply go.mod:221]: I can change the replace to point to my fork for now, as the branch exists on my fork I don't think go mod can pull it from the aquasecurity while the PR is not merged.

Comment (knqyf263) [reply go.mod:221]: Please try
```
replace github.com/aquasecurity/trivy-kubernetes =>  https://github.com/josedonizetti/trivy-kubernetes@ fcfb51c0e4ebb6941377f1a4e732de54a318847f
```

**Comment by josedonizetti** [comment]
@chen-keinan @knqyf263 Can you review? This is a first version of the command, limited features, so we can release something working for kubecon, then I follow up with other PRs to improve things. 

Things TODO:

1 - Add more formats, for now the only format is json.
2 - Better performance, for now we are doing one scan per time, no concurrency.
3 - Better unit tests
3 - Some flags are commented out to be implemented later.
4 - Scan secrets
5 - Use memory FS instead of writing resources to files
6 - Reports (summary, others?)

Let me know what you think is a blocker on your review, and what I can fix on a following PR please.

File: pkg/commands/artifact/k8s.go
Comment (chen-keinan) [inline pkg/commands/artifact/k8s.go:70]: This condition is redundant as there is a check already in `trivy-kubernetes` :
https://github.com/aquasecurity/trivy-kubernetes/blob/14575687ba8c20fc5281154b944f268e105637ab/pkg/trivyk8s/trivyk8s.go#L54

File: pkg/commands/artifact/k8s.go
Comment (chen-keinan) [inline pkg/commands/artifact/k8s.go:238]: file handler remains open if it fails on `yaml.Marshal`

File: pkg/commands/artifact/k8s.go
Comment (chen-keinan) [inline pkg/commands/artifact/k8s.go:243]: file handler remains open if it fails on `file.Write`

File: pkg/commands/artifact/k8s.go
Comment (chen-keinan) [inline pkg/commands/artifact/k8s.go:206]: Resource config file removal should be deferred right after its creation (L195)  , as if `k8sScan` fails, the file will remains on the disk

**Review by chen-keinan — CHANGES_REQUESTED** [review]

**Review by chen-keinan — APPROVED** [review]

File: docs/docs/kubernetes/scanning.md
Comment (knqyf263) [inline docs/docs/kubernetes/scanning.md:30]: This is an example. We don't have to put the entire JSON. Can we shorten it?

**Review by knqyf263 — COMMENTED** [review]
You need to update mkdocs.yml as well. You can see how to build the doc locally here.
https://aquasecurity.github.io/trivy/v0.27.1/community/contribute/pr/#documentation

File: pkg/commands/artifact/k8s.go
Comment (knqyf263) [inline pkg/commands/artifact/k8s.go:191]: This verb can be used only in 'fmt.Errorf' calls (%w).

```suggestion
			log.Logger.Errorf("failed to delete temp file %s: %s", fileName, err)
```

File: pkg/commands/artifact/k8s.go
Comment (knqyf263) [inline pkg/commands/artifact/k8s.go:196]: This `defer` should be put after the error handling. If it fails to create a temp file, it also fails to remove the file.
```suggestion
	if err != nil {
		return types.Report{}, xerrors.Errorf("scan error: %w", err)
	}
	defer func() {
		if err := os.Remove(fileName); err != nil {
			log.Logger.Errorf("failed to delete temp file %s:%w:", fileName, err)
		}
	}()
```

File: go.mod
Comment (knqyf263) [inline go.mod:243]: Looks like PRs are already in the upstream (trivy-kubernetes). Do we still need the `replace`?

File: pkg/commands/app.go
Comment (knqyf263) [inline pkg/commands/app.go:788]: I think "scan kubernetes images and configurations" or "scan kubernetes vulnerabilities and misconfigurations" are better since `image` is an artifact and `misconfigurations` is a security issue.

File: pkg/commands/artifact/k8s.go
Comment (knqyf263) [inline pkg/commands/artifact/k8s.go:156]: The version 0 looks weird. Is this because this feature is still experimental?

File: docs/docs/kubernetes/scanning.md
Comment (knqyf263) [inline docs/docs/kubernetes/scanning.md:2]: Can we mention this is experimental here as below?
http://localhost:8000/trivy/docs/vulnerability/examples/filter/#by-open-policy-agent

File: pkg/commands/artifact/k8s.go
Comment (knqyf263) [inline pkg/commands/artifact/k8s.go:126]: Any reason this is a debug message? I understand the error will be included in the report, but I feel like we should notify it in terminal.

File: pkg/commands/artifact/k8s.go
Comment (knqyf263) [inline pkg/commands/artifact/k8s.go:272]: Looks like this issue is remaining.
https://github.com/aquasecurity/trivy/pull/2065#discussion_r869539653

File: go.mod
Comment (josedonizetti) [reply go.mod:243]: Yes, I'm waiting for a last merge -> https://github.com/aquasecurity/trivy-kubernetes/pull/12

File: pkg/commands/artifact/k8s.go
Comment (josedonizetti) [reply pkg/commands/artifact/k8s.go:156]: That was the idea, but I can change it to 1 if you think it makes more sense.

File: go.mod
Comment (knqyf263) [reply go.mod:243]: OK, let's wait for it then.

File: pkg/commands/artifact/k8s.go
Comment (knqyf263) [reply pkg/commands/artifact/k8s.go:156]: It is ok then. I agree the schema is not fixed yet. I'm sure we will fix the schema soon.

File: pkg/commands/artifact/k8s.go
Comment (josedonizetti) [reply pkg/commands/artifact/k8s.go:126]: I asked a few friends to try it on their cluster with several resources (500-1000 resources), several resources were deployed with private images, and `trivy` was not configured to access the registry, the output becomes terrible, lots and lots of error logs, seems like everything is not working, and they cancelled the command, without waiting to generate a report , even to check at least the misconfigurations as the images were not reachable.

Comment (knqyf263) [reply pkg/commands/artifact/k8s.go:126]: Interesting. We probably should notify users if it is an authentication error like "To scan private images, please consider setting appropriate credentials." Let's keep it in our mind as a future enhancement.

File: docs/docs/kubernetes/scanning.md
Comment (owenrumney) [inline docs/docs/kubernetes/scanning.md:4]:
```suggestion
This feature might change without preserving backwards compatibility.
```

File: docs/docs/kubernetes/scanning.md
Comment (owenrumney) [inline docs/docs/kubernetes/scanning.md:14]:
```suggestion
Scan a specific namespace:
```

File: pkg/commands/app.go
Comment (owenrumney) [inline pkg/commands/app.go:214]:
```suggestion
		Usage:   "specify a specific namespace to scan",
```

File: pkg/commands/app.go
Comment (owenrumney) [inline pkg/commands/app.go:793]: Remove commented code and readd when required?

**Review by owenrumney — COMMENTED** [review]
These are all just suggestions - mostly docs

**Review by knqyf263 — APPROVED** [review]

**Review by owenrumney — APPROVED** [review]

**Pull request merged**
